packages/world/ts/actions/delegation.ts

```diff
@@ -0,0 +1,56 @@
+import type { WalletClient, Transport, Chain, Account, Hex, WalletActions, WriteContractReturnType } from "viem";
+import { getAction, getAbiItem, formatAbiItem, toFunctionSelector, type GetAbiItemParameters } from "viem/utils";
+import { writeContract } from "viem/actions";
+import { encodeSystemCallFrom, type SystemCallFrom } from "../encodeSystemCallFrom";
+
+type DelegationParameters = {
+  worldAddress: Hex;
+  delegatorAddress: Hex;
+  getSystemId: (functionSelector: Hex) => Hex;
+};
+
+// By extending clients with this function after delegation, the delegation automatically applies
+// to the World contract writes, meaning these calls are made on behalf of the delegator.
+export function delegation<TChain extends Chain, TAccount extends Account>({
+  worldAddress,
+  delegatorAddress,
+  getSystemId,
+}: DelegationParameters): (
+  client: WalletClient<Transport, TChain, TAccount>,
+) => Pick<WalletActions<TChain, TAccount>, "writeContract"> {
+  return (client) => ({
+    // Applies to: `client.writeContract`, `getContract(client, ...).write`
+    writeContract: (originalArgs): Promise<WriteContractReturnType> => {
+      // Skip if the contract isn't the World.
+      if (originalArgs.address !== worldAddress) {
+        return getAction(client, writeContract, "writeContract")(originalArgs);
+      }
+
+      // Ensured not to be `undefined` because of the `writeContract` args type.
+      const functionAbiItem = getAbiItem({
+        abi: originalArgs.abi,
+        name: originalArgs.functionName,
+        args: originalArgs.args,
+      } as unknown as GetAbiItemParameters)!;
+
+      // `callFrom` requires `systemId`.
+      const functionSelector = toFunctionSelector(formatAbiItem(functionAbiItem));
+      const systemId = getSystemId(functionSelector);
```

Comment:
I think we might be missing some nuance around world function selectors vs system function selectors here.

When we call a namespaced system, the `originalArgs.functionName` will be something like `namespace__functionName` and the ABI will likely be the `IWorld` ABI (all of the world functions).

But when we translate that to a system function call, we need to pass in the system ID + system function selector (i.e. just `functionName` without the namespace, since the namespace is part of the system ID).

Maybe `getSystemId` needs to be something like `worldFunctionToSystemFunction` or `getSystemCall` or similar? It takes in a world function selector and returns both the system ID and system function selector. The args will be the same, but we may have to manually encode the args instead of using `encodeSystemCallFrom`.

Reply:
I was misunderstanding this point. As you mentioned, it only worked for root namespace systems.